### DISCUSSION TITLE
recovery tag data

### DISCUSSION THREAD

**Discussion by sourcema** [alefragnani/vscode-bookmarks#640] [Q&A] [answered]
Hello, I use the bookmark under the window to store the tags of the remote code, but now the ip address of the remote server has changed, so my bookmarks are gone. How can I restore the data, or go there to find the data information stored before? Thanks.

**Comment by sourcema** [comment]
tag description also save in state.vscdb? i did't found it.

**Comment by sourcema** [comment]
I found that the C:\Users\Administrator\AppData\Roaming\Code\User\workspaceStorage directory is a single temporary file. Is there a file that stores all the data? Does the plug-in perform data recovery based on so many db files?

**Comment by alefragnani** [comment, marked as answer]
Hi @sourcema ,

The extension stores the bookmarks _per workspace_ (which is a VS Code concept) and when you rename / move it, VS Code will understand it as a new workspace. For remotes, it't s the same concept, and if the IP of the remote you are connected (via SSH for instance) changes, it will be recognized as a new workspace as well. That's the reason why aren't seeing the bookmarks from the original IP.

In the old days, the bookmarks were stored inside VS Code itself (Local Storage) and were accessible using DevTools, but it's been a while the data is stored locally:

```ts
var appdata = process.env.APPDATA || (process.platform === "darwin" ? process.env.HOME + "/Library/Application Support" : "/var/local");

appdata += "/Code/User/workspaceState";
```

Inside the `workspaceState` folder, you will see a lot of _guid like_ folders, and inside each one, the workspace state, which is a SQLite database with all the data VS Code stores for that particular workspace. The main problem is to identify which of those _guid like_ folders is related to your workspace. And, depending on how you use Remote extension, these files may be _on the remote_

Unfortunately, there is no solution to recover the data using the extension. But, there are alternatives to not happen anymore

1. Stores the bookmarks within the workspace, using the `bookmarks.saveBookmarksInProject` setting
2. The [Save bookmarks as global](https://github.com/alefragnani/vscode-bookmarks/issues/287) feature request, which still has no ETA

Hope this helps

**Reply by sourcema** [reply]
I got it, thanks for your answer.
